lib/bolt/module_installer.rb

```diff
@@ -13,45 +17,49 @@ def initialize(outputter, pal)
-      require 'bolt/puppetfile'
-
-      @outputter.print_message("Adding module #{name} to project\n\n")
-
-      # If the project configuration file already includes this module,
-      # exit early.
-      puppetfile  = Bolt::Puppetfile.new(modules)
-      new_module  = Bolt::Puppetfile::Module.from_hash('name' => name)
-
-      if puppetfile.modules.include?(new_module)
-        @outputter.print_action_step(
-          "Project configuration file #{config_path} already includes module #{new_module}. Nothing to do."
+    def add(name, specs, puppetfile_path, moduledir, config_path)
+      specs = Specs.new(specs)
+
+      # Exit early if project config already includes a spec with this name.
+      if specs.include?(name)
+        @outputter.print_message(
+          "Project configuration file #{config_path} already includes specification with name "\
+          "#{name}. Nothing to do."
         )
         return true
       end
 
+      @outputter.print_message("Adding module #{name} to project\n\n")
+
+      # Create a Puppetfile object. If there isn't an existing Puppetfile,
+      # this won't include any resolved modules.
+      existing_puppetfile = Puppetfile.parse(puppetfile_path)
+      existing_specs      = Specs.from_puppetfile(existing_puppetfile)
+
       # If the Puppetfile exists, make sure it's managed by Bolt.
-      if puppetfile_path.exist?
-        assert_managed_puppetfile(puppetfile, puppetfile_path)
-        existing = Bolt::Puppetfile.parse(puppetfile_path)
-      else
-        existing = Bolt::Puppetfile.new
-      end
+      existing_puppetfile.assert_satisfies(specs) if puppetfile_path.exist?
+
+      # Resolve module dependencies. Attempt to first resolve with resolved
+      # modules. If that fails, fall back to resolving all specs.
+      @outputter.print_action_step("Resolving module dependencies, this may take a moment")
 
-      # Create a Puppetfile object that includes the new module and its
-      # dependencies. We error early here so we don't add the new module to the
-      # project config or modify the Puppetfile.
-      puppetfile = add_new_module_to_puppetfile(new_module, modules, puppetfile_path)
+      begin
+        existing_specs.add_specs('name' => name)
+        puppetfile = Resolver.new.resolve(existing_specs)
+      rescue Bolt::Error
+        specs.add_specs('name' => name)
+        puppetfile = Resolver.new.resolve(specs)
+      end
 
       # Display the diff between the existing Puppetfile and the new Puppetfile.
-      print_puppetfile_diff(existing, puppetfile)
+      print_puppetfile_diff(existing_puppetfile, puppetfile)
 
       # Add the module to the project configuration.
       @outputter.print_action_step("Updating project configuration file at #{config_path}")
 
       data = Bolt::Util.read_yaml_hash(config_path, 'project')
       data['modules'] ||= []
-      data['modules'] <<  { 'name' => new_module.title }
+      data['modules'] <<  name
```

Comment:
Isn't `data['modules']` the same as the original `specs` here? We shouldn't need to read the project file twice. Separately, we should rename `specs` to `project_specs`, and I think `existing_specs` to `puppetfile_specs`.

Reply:
We can't update `bolt-project.yaml` without loading all of it and modifying the `modules` key...

Reply:
But haven't we already read it when we loaded the project? `specs` here should be the same as `data[modules]`

Reply:
Oh I see what you're saying - I mean, we're passing so many things from the project class here already. We could pass data instead of `specs` and get specs from that. That'd also make it more clear where specs is coming from.

Reply:
It's a minor thing, but `data` doesn't necessarily have everything that's in the file since we reject transport config.

Also not super keen on passing all of the project config. Reading the file a second time seems fairly minor.